File: java/socketd/src/main/java/org/noear/socketd/SocketD.java
package org.noear.socketd;

import org.noear.socketd.cluster.ClusterClient;
import org.noear.socketd.transport.client.ClientProvider;
import org.noear.socketd.transport.core.Asserts;
import org.noear.socketd.transport.server.ServerProvider;
import org.noear.socketd.transport.client.Client;
import org.noear.socketd.transport.client.ClientConfig;
import org.noear.socketd.transport.server.Server;
import org.noear.socketd.transport.server.ServerConfig;
import org.noear.socketd.utils.ProviderUtils;

import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.ServiceLoader;

/**
 * @author noear
 * @since 2.0
 */
public class SocketD {

    /**
     * 框架版本号
     */
    public static String version() {
        return "2.5.11";
    }

    /**
     * 协议名
     */
    public static String protocolName() {
        return "Socket.D";
    }

    /**
     * 协议版本号
     */
    public static String protocolVersion() {
        return "1.0";
    }

    /**
     * 客户端提供者
     */
    private static Map<String, ClientProvider> clientProviderMap = new HashMap<>();

    /**
     * 服务端提供者
     */
    private static Map<String, ServerProvider> serverProviderMap = new HashMap<>();

    /**
     * 加载 spi
     * */
    static {
        ServiceLoader.load(ClientProvider.class).iterator().forEachRemaining(clientProvider -> {
            registerClientProvider(clientProvider);
        });

        ServiceLoader.load(ServerProvider.class).iterator().forEachRemaining(serverProvider -> {
            registerServerProvider(serverProvider);
        });

        ProviderUtils.autoWeakLoad();
    }

    /**
     * 手动注册客户端提供者
     */
    public static void registerClientProvider(ClientProvider clientProvider) {
        for (String s : clientProvider.schemas()) {
            clientProviderMap.put(s, clientProvider);
        }
    }

    /**
     * 手动注册服务端提供者
     */
    public static void registerServerProvider(ServerProvider serverProvider) {
        for (String s : serverProvider.schemas()) {
            serverProviderMap.put(s, serverProvider);
        }
    }

    /**
     * 创建服务端
     */
    public static Server createServer(String schema) {
        Server server = createServerOrNull(schema);
        if (server == null) {
            throw new IllegalStateException("No socketd server providers were found: " + schema);
        } else {
            return server;
        }
    }

    /**
     * 创建服务端，如果没有则为 null
     */
    public static Server createServerOrNull(String schema) {
        Asserts.assertNull("schema", schema);

        ServerProvider factory = serverProviderMap.get(schema);
        if (factory == null) {
            return null;
        } else {
            return factory.createServer(new ServerConfig(schema));
        }
    }

    /**
     * 创建客户端（支持 url 自动识别）
     *
     * @param serverUrl 服务器地址
     */
    public static Client createClient(String serverUrl) {
        Client client = createClientOrNull(serverUrl);
        if (client == null) {
            throw new IllegalStateException("No socketd client providers were found: " + serverUrl);
        } else {
            return client;
        }
    }

    /**
     * 创建客户端（支持 url 自动识别），如果没有则为 null
     *
     * @param serverUrl 服务器地址
     */
    public static Client createClientOrNull(String serverUrl) {
        Asserts.assertNull("serverUrl", serverUrl);

        int idx = serverUrl.indexOf("://");
        if (idx < 2) {
            throw new IllegalArgumentException("The serverUrl invalid: " + serverUrl);
        }

        String schema = serverUrl.substring(0, idx);

        ClientProvider factory = clientProviderMap.get(schema);
        if (factory == null) {
            return null;
        } else {
            ClientConfig clientConfig = new ClientConfig(serverUrl);
            return factory.createClient(clientConfig);
        }
    }

    /**
     * 创建集群客户端
     *
     * @param serverUrls 服务端地址
     */
    public static ClusterClient createClusterClient(String... serverUrls) {
        return new ClusterClient(serverUrls);
    }

    /**
     * 创建集群客户端
     *
     * @param serverUrls 服务端地址
     */
    public static ClusterClient createClusterClient(List<String> serverUrls) {
        return new ClusterClient(serverUrls.toArray(new String[serverUrls.size()]));
    }
}

File: python/socketd/socketd/SocketD.py
from typing import Dict

from socketd.cluster.ClusterClient import ClusterClient
from socketd.transport.client.ClientConfig import ClientConfig
from socketd.transport.client.Client import Client
from socketd.transport.client.ClientProvider import ClientProvider
from socketd.transport.core.Asserts import Asserts
from socketd.transport.server.Server import Server
from socketd.transport.server.ServerConfig import ServerConfig
from socketd.transport.server.ServerProvider import ServerProvider

from socketd_websocket.WsAioProvider import WsAioProvider
from socketd_aio_tcp.TcpAioProvider import TcpAioProvider


def version() -> str:
    return "2.5.11"

def protocol_name() -> str:
    return "Socket.D"

def protocol_version() -> str:
    return "1.0"


__client_provider_map: Dict[str, ClientProvider] = {}
__server_provider_map: Dict[str, ServerProvider] = {}

def register_client_provider(clientProvider: ClientProvider):
    for s in clientProvider.schema():
        __client_provider_map[s] = clientProvider

def register_server_provider(serverProvider: ServerProvider):
    for s in serverProvider.schema():
        __server_provider_map[s] = serverProvider

def create_server(schemaOrConfig: str | ServerConfig) -> Server:
    Asserts.assert_null("schemaOrConfig", schemaOrConfig)
    config:ServerConfig

    if isinstance(schemaOrConfig, str):
        config = ServerConfig(schemaOrConfig)
    else:
        config = schemaOrConfig

    server = create_server_or_null(config)

    if server is None:
        raise RuntimeError(f"No socketd server providers were found: {config.get_schema()}")
    else:
        return server

def create_server_or_null(config: ServerConfig) -> Server:
    provider = __server_provider_map.get(config.get_schema())

    if provider is None:
        return None
    else:
        return provider.create_server(config)


def create_client(urlOrConfig: str | ClientConfig) -> Client:
    Asserts.assert_null("urlOrConfig", urlOrConfig)
    config:ClientConfig

    if isinstance(urlOrConfig, str):
        config = ClientConfig(urlOrConfig)
    else:
        config = urlOrConfig

    client = create_client_or_null(config)

    if client is None:
        raise RuntimeError(f"No socketd client providers were found: {config.get_schema()}")
    else:
        return client

def create_client_or_null(config: ClientConfig) -> Client | None:
    Asserts.assert_null("config", config)

    provider = __client_provider_map.get(config.get_schema())
    if provider is None:
        return None
    else:
        return provider.create_client(config)


def create_cluster_client(*urls) -> Client:
    return ClusterClient(urls)


# Initialize the client and server provider maps
for p in [WsAioProvider(), TcpAioProvider()]:
    register_client_provider(p)
    register_server_provider(p)
